x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/FrozenEngine.java

```diff
@@ -232,6 +238,45 @@ static LazyDirectoryReader unwrapLazyReader(DirectoryReader reader) {
         return null;
     }
 
+    /*
+     * We register this listener for a frozen index that will
+     *  1. reset the reader every time the search context is validated which happens when the context is looked up ie. on a fetch phase
+     *  etc.
+     *  2. register a releasable resource that is cleaned after each phase that releases the reader for this searcher
+     */
+    public static class ReacquireEngineSearcherListener implements SearchOperationListener {
+
+        @Override
+        public void validateSearchContext(SearchContext context, TransportRequest transportRequest) {
+            Searcher engineSearcher = context.searcher().getEngineSearcher();
+            LazyDirectoryReader lazyDirectoryReader = unwrapLazyReader(engineSearcher.getDirectoryReader());
+            if (lazyDirectoryReader != null) {
+                try {
+                    lazyDirectoryReader.reset();
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
+            // also register a release resource in this case if we have multiple roundtrips like in DFS
+            onNewContext(context);
```

Comment:
Nit: maybe only call this if we have a lazyDirectoryReader and the rest was successful?  slightly easier to follow the logic.